tfhe/src/safe_serialization.rs

```diff
@@ -0,0 +1,149 @@
+use bincode::Options;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+
+const VERSION: &str = "0.3.0";
+
+const VERSION_LENGTH_LIMIT: u64 = 100;
+
+const TYPE_LENGTH_LIMIT: u64 = 1000;
+
+pub fn safe_serialize<T: Serialize>(
+    object: &T,
+    mut writer: impl std::io::Write,
+    serialized_size_limit: u64,
+) -> bincode::Result<()> {
+    let options = bincode::DefaultOptions::new()
+        .with_fixint_encoding()
+        .with_limit(0);
+
+    options
+        .with_limit(VERSION_LENGTH_LIMIT)
+        .serialize_into::<_, String>(&mut writer, &VERSION.to_owned())?;
+
+    options
+        .with_limit(TYPE_LENGTH_LIMIT)
+        .serialize_into::<_, String>(&mut writer, &std::any::type_name::<T>().to_owned())?;
+
+    options
+        .with_limit(serialized_size_limit)
+        .serialize_into(&mut writer, object)?;
+
+    Ok(())
+}
+
+pub fn safe_deserialize<T: DeserializeOwned>(
+    mut reader: impl std::io::Read,
+    serialized_size_limit: u64,
+) -> Result<T, String> {
+    let options = bincode::DefaultOptions::new()
+        .with_fixint_encoding()
+        .with_limit(0);
```

Comment:
set an upper limit at all times for the deserialize to avoid OOM, if we forget to call a with limit we may have some problems